internal/app/machined/pkg/controllers/runtime/oom.go

```diff
@@ -0,0 +1,208 @@
+			continue
+		}
+
+		value, ok := spans[pressureSpan]
+		if !ok {
+			fmt.Println("cannot find memory pressure span:", pressureSpan)
+			continue
+		}
+
+		if !value.IsSet || value.IsMax {
+			continue
+		}
+
+		val, err := strconv.ParseFloat(value.String(), 32)
+		if err != nil {
+			fmt.Println("cannot parse memory pressure:", pressureSpan, err)
+			continue
+		}
+		fmt.Println("monitoring", value.String(), val, err)
+
+		if val > psiThresh && time.Since(ctrl.ActionTriggered) > cooldownTimeout {
+			ctrl.ActionTriggered = time.Now()
+			ctrl.OomAction()
+		}
+	}
+}
+
+// OomAction
+func (ctrl *OOMController) OomAction() {
+	fmt.Println("OOM action!")
+
+	ranking := []oomRankedCgroup{}
+
+	for _, cg := range []struct {
+		dir   string
+		class int
+	}{
+		{"kubepods/besteffort", OomCgroupClassBesteffort},
+		{"kubepods/burstable", OomCgroupClassBurstable},
+		{"kubepods/guaranteed", OomCgroupClassGuaranteed},
+		{"podruntime", OomCgroupClassPodruntime},
+		{"system", OomCgroupClassSystem},
+	} {
+		entries, err := os.ReadDir(filepath.Join(constants.CgroupMountPath, cg.dir))
+		if err != nil {
+			fmt.Println("cannot list cgroup members", cg.dir, err)
+			continue
+		}
+
+		for _, leaf := range entries {
+			if !leaf.IsDir() {
+				continue
+			}
+
+			leafDir := filepath.Join(constants.CgroupMountPath, cg.dir, leaf.Name())
+
+			node := cgroups.Node{}
+
+			for _, prop := range []string{"memory.current", "memory.peak", "memory.max"} {
+				err := cgroups.ReadCgroupfsProperty(&node, leafDir, prop)
+				if err != nil {
+					fmt.Println("cannot read property for cgroup", leafDir, prop, err)
+					continue
+				}
+			}
+
+			ranking = append(ranking, oomRankedCgroup{
+				Path:          leafDir,
+				Class:         cg.class,
+				MemoryCurrent: node.MemoryCurrent.Val,
+				MemoryPeak:    node.MemoryPeak.Val,
+				MemoryMax:     node.MemoryMax.Val,
+			})
+		}
+	}
+
+	if len(ranking) == 0 {
+		return
+	}
+
+	// Sort processes by the following hierarchy:
+	// First, sort by high-level group:
+	//     kubepods (workloads)
+	//     podruntime (CRI, kubelet, etcd)
+	//     runtime (core containerd, system services)
+	//     init
+	// Second, inside kubepods we have QoS groups:
+	//     first priority: BestEffort
+	//     second: Burstable
+	//     last: Guaranteed
+	// Third, look into other attributes, e.g. OOM score.
+	// Fourth, look into memory max - memory current (if memory max is set).
+	//
+	// Sort to make the most prioritized to OOM-kill cgroup to the first place
+	//
+	// TODO: implement CEL or other configurable ranking method
+	sort.Slice(ranking, func(i int, j int) bool {
+		a, b := ranking[i], ranking[j]
+		if a.Class == b.Class {
+			return a.MemoryCurrent > b.MemoryCurrent
+		}
+
+		return a.Class < b.Class
+	})
+
+	fmt.Println(ranking)
+	fmt.Println("SENDING SIGKILL TO CGROUP", ranking[0])
+
+	os.WriteFile(filepath.Join(ranking[0].Path, "cgroup.kill"), []byte{'1'}, 0o644)
```

Comment:
I wonder if we should also create a resource(s) to store something like ten last killed cgroup names as the output of the controller.

We can also store there PSI at the moment we killed, the CGroup properties - why it was picked, etc.